packages/flutter/lib/src/material/focusable.dart

```diff
@@ -0,0 +1,297 @@
+
+import 'theme.dart';
+
+typedef FocusableOnKeyCallback = bool Function(FocusableNode node, RawKeyEvent event);
+
+/// A widget that manages a [FocusableNode] and allows specification of the
+/// focus order.
+///
+/// The node can be either a scope node or a leaf node. Scope nodes provide a
+/// scope for their children, using the focus traversal policy defined by the
+/// [DefaultFocusTraversal] widget above them to traverse their children.
+///
+/// It provides [onFocusChange] as a way to be notified when the focus is given
+/// to or removed from this widget, and allows specification of a [focusedDecoration]
+/// to be shown when its [child] has focus.
+///
+/// The [onKey] argument allows specification of a key even handler that should
+/// be invoked when this node or one of its children has focus.
+class Focusable extends StatefulWidget {
+  /// Creates a widget that manages a [FocusableNode]
+  ///
+  /// The [child] arguments is required and must not be null.
+  ///
+  /// The [isScope], [autofocus], and [showDecorations] arguments must not be
+  /// null.
+  const Focusable({
+    Key key,
+    @required this.child,
+    this.isScope = false,
+    this.autofocus = false,
+    this.showDecorations = true,
+    this.focusedDecoration,
+    this.unfocusedDecoration,
+    this.onFocusChange,
+    this.onKey,
+    this.debugLabel,
+  })  : assert(child != null),
+        assert(isScope != null),
+        assert(autofocus != null),
+        assert(showDecorations != null),
+        super(key: key);
+
+  /// A debug label for this widget.
+  final String debugLabel;
+
+  /// The child widget of this [Focusable].
+  ///
+  /// {@macro flutter.widgets.child}
+  final Widget child;
+
+  /// Handler for keys pressed when this object or one of its children has
+  /// focus.
+  ///
+  /// Key events are first given to the leaf nodes, and if they don't handle
+  /// them, then to each node up the widget hierarchy. If they reach the root of
+  /// the hierarchy, they are discarded.
+  final FocusableOnKeyCallback onKey;
+
+  /// Handler called when the focus of this focusable changes.
+  ///
+  /// Called with true if this focusable gains focus, and false if it loses
+  /// focus.
+  final ValueChanged<bool> onFocusChange;
+
+  /// True if this [Focusable] serves as a scope for other [Focusable]s.
+  ///
+  /// This means that it remembers the last focusable that was focused within
+  /// its descendants, and can move that focus to the next/previous node, or a
+  /// node in a particular direction when the [FocusableNode.nextFocus],
+  /// [FocusableNode.previousFocus], or [FocusableNode.focusInDirection] are
+  /// called on a Focusable node that is a child of this scope, or the node
+  /// owned by this scope node.
+  ///
+  /// The selection process of the node to move to is determined by the node
+  /// traversal policy specified by the nearest enclosing
+  /// [DefaultFocusTraversal] widget.
+  final bool isScope;
```

Comment:
I encourage you to see if you can implement this using inheritance rather than a flag. See if you can change the places where you test `isScope` and instead rely on virtual methods that behave one way or the other. Failing that, rather than having it be a constructor argument, try to use two constructors.

Reply:
Yeah, I've already gone back and forth a couple of times on this, and after determining that I really do need a "scope" (I was hoping for a while that I could do it all with just a hierarchy with one kind of node), I agree with you.  I also think I will do the same for Focusable and make a FocusableScope that is a Focusable that creates a scope.  I also want to look one more time to see if I can replace FocusScope and the FocusNode guts with these guts and just add API, but the changes are still extensive, and I don't have a lot of confidence that it would be a smooth replacement for users of FocusNode (and especially FocusScope).